api/include/opentelemetry/trace/span.h

```diff
@@ -1,38 +1,112 @@
+  // must be provided.
   //
-  // Timestamps from the steady clock can be used to most accurately measure a span's
-  // duration, while timestamps from the system clock can be used to most accurately place a span's
-  // time point relative to other spans collected across a distributed system.
+  // Timestamps from the steady clock can be used to most accurately measure a Span's
+  // duration, while timestamps from the system clock can be used to most accurately place a Span's
+  // time point relative to other Spans collected across a distributed system.
   core::SystemTimestamp start_system_time;
   core::SteadyTimestamp start_steady_time;
+
+  // TODO:
+  // Span(Context?) parent;
+  // SpanContext remote_parent;
+  // Links
+  // Attributes
+  SpanKind kind = SpanKind::kInternal;
 };
 
 class Tracer;
 
 /**
- * A span represents a single operation within a trace.
+ * A Span represents a single operation within a Trace.
+ *
+ * Spans *must* be ended by calling End(). Merely deleting a Span does not end
+ * it.
  */
 class Span
 {
 public:
+  // Note that Spans should be created using the Tracer class. Please refer to
+  // tracer.h for documentation.
+  Span()          = default;
   virtual ~Span() = default;
 
+  // Not copiable or movable.
+  Span(const Span &) = delete;
+  Span(Span &&)      = delete;
+  Span &operator=(const Span &) = delete;
+  Span &operator=(Span &&) = delete;
+
+  // TODO
+  // Sets an attribute on the Span. If the Span previously contained a mapping for
+  // the key, the old value is replaced.
+  //
+  // If an empty string is used as the value, the attribute will be silently
+  // dropped. Note: this behavior could change in the future.
+  // virtual void SetAttribute(nostd::string_view key, AttributeValue&& value) = 0;
+
+  // Adds an event to the Span.
+  virtual void AddEvent(nostd::string_view name) = 0;
+
+  // Adds an event to the Span, with a custom timestamp.
+  virtual void AddEvent(nostd::string_view name, core::SystemTimestamp timestamp) = 0;
+  virtual void AddEvent(nostd::string_view name, core::SteadyTimestamp timestamp) = 0;
+
+  // TODO
+  // Adds an event to the Span, with a custom timestamp, and attributes.
+  // virtual void AddEvent(nostd::string_view name, core::SteadyTimestamp
+  // timestamp, nostd::span<std::pair<nostd::string_view name, AttributeValue&&
+  // value>> attributes);
+
+  // TODO
+  // Sets the status of the span. The default status is OK. Only the value of the last call will be
+  // recorded, and implementations are free to ignore previous calls.
+  // virtual void SetStatus(Status status) = 0;
+
+  // Updates the name of the Span. If used, this will override the name provided
+  // during construction.
+  virtual void UpdateName(nostd::string_view name) = 0;
+
+  // Mark the end of the Span. Only the timing of the first end call for a given {@code Span} will
+  // be recorded, and implementations are free to ignore all further calls.
+  virtual void End() = 0;
```

Comment:
Can we make these noexcept, especially something like End which is likely to be called from a destructor?

If any of these fail, I'd expect the tracer to turn them into noops.

Reply:
Done.

Considering "API can be used without exceptions", should every function in the API be marked noexcept explicitly?